fail-safely.md

```diff
@@ -0,0 +1,100 @@
+# Fail Safely
+
+Go doesn't have enums. (at least at the time of this writing and I have no idea if there are plans of adding them.)
+
+I've typically seen related values expressed as a set of constants.
+
+```go
+const (
+	Active  = "A"
+	Deleted = "D"
+	Pending = "P"
+)
+```
+
+or using iota. (no, stop it)
+
+```go
+const (
+  Active = iota + 1
+  Deleted
+  Pending
+)
+```
+
+However, in code, I may find something like:
+
+```go
+if user.State == Deleted {
+  return someErr
+}
+```
+
+which is doomed to fail open. What happens when a new state is added? Yep, it's _implicitly_ allowed. Congratulations, the new `Purged` state is now allowed where `Deleted` was not.
+
+We can make it a little better by describing what we want to continue, and negating it. See also [Happy Path Expressions](./negated-happy-path-form.md).
+
+```go
+if user.State != Active {
+  return someErr
+}
+```
+
+While this is _better_, we're still _implicitly_ denying any other type, but we're talking about a known set of values here. The best thing to do would be to exhaustively handle every related state, _explicitly_ allowing or denying each one, and by doing so, protecting against that future new state with a default deny.
+
+```go
+switch user.State {
+  case Active, Pending:
```

Comment:
Another technique I can imagine is having defined groups or categories of types that are defined once and referenced in relevant code paths.

Though even with that, the logic should fail closed outside of that defined group.

Reply:
I did consider and play with this idea, but it feels like a DRYing effort, and conditional checks are better thought of as purpose-built for their context. While adding a layer of abstraction is possible, in most cases it's unnecessary.

The bit of exploration I did felt like I was abstracting a switch statement and/or a map. I'd be interested in seeing an example of what you're imagining if you have time. (no worries if not ❤️)

Reply:
> While adding a layer of abstraction is possible, in most cases it's unnecessary.

Totally! In some cases it's helpful, depends on the context. I agree it's best to be cautious about reaching for abstraction too early.